File: examples/02-reliability/go/cancellation_scope.go
package reliability

import (
	"time"

	"go.temporal.io/sdk/temporal"
	"go.temporal.io/sdk/workflow"
)

// Start a long-running export under a cancellable child context, then race it
// against a deadline timer with a Selector. When the deadline wins we call
// cancel(): that propagates to the Activity on its next heartbeat (its ctx.Err
// becomes canceled) so it can clean up partial work.
func ExportWithDeadline(ctx workflow.Context, tableName string, deadline time.Duration) (string, error) {
	actCtx := workflow.WithActivityOptions(ctx, workflow.ActivityOptions{
		StartToCloseTimeout: 2 * time.Hour,
		HeartbeatTimeout:    30 * time.Second,
	})
	childCtx, cancel := workflow.WithCancel(actCtx)

	exportF := workflow.ExecuteActivity(childCtx, "ExportLargeTable", tableName)
	timer := workflow.NewTimer(ctx, deadline)

	var result string
	var finishedInTime bool
	workflow.NewSelector(ctx).
		AddFuture(exportF, func(f workflow.Future) {
			finishedInTime = true
			_ = f.Get(ctx, &result)
		}).
		AddFuture(timer, func(workflow.Future) { cancel() }).
		Select(ctx)

	if !finishedInTime {
		return "", temporal.NewApplicationError("export timed out", "ExportTimeout")
	}
	return result, nil
}

File: examples/02-reliability/python/cancellation_scope.py
import asyncio
from datetime import timedelta

from temporalio import workflow
from temporalio.exceptions import ApplicationError

# `export_large_table` is the heartbeating Activity from heartbeat_long_activity.py.


@workflow.defn
class CancellableExportWorkflow:
    # Start a long-running export as a cancellable task, then race it against a
    # deadline. asyncio.wait_for cancels the task when the timeout fires; that
    # cancellation reaches the Activity on its next heartbeat (raising
    # CancelledError there) so it can clean up partial work.
    @workflow.run
    async def export_with_deadline(self, table_name: str, deadline_secs: int) -> str:
        export = workflow.execute_activity(
            export_large_table,
            table_name,
            start_to_close_timeout=timedelta(hours=2),
            heartbeat_timeout=timedelta(seconds=30),
        )
        try:
            return await asyncio.wait_for(export, timeout=deadline_secs)
        except asyncio.TimeoutError:
            raise ApplicationError("export timed out", type="ExportTimeout")
